plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarMetadata.java

```diff
@@ -0,0 +1,438 @@
+                        public TopicName load(SchemaTableName schemaTableName) throws Exception
+                        {
+                            return getMatchedPulsarTopic(schemaTableName);
+                        }
+                    });
+
+    @Inject
+    public PulsarMetadata(
+            PulsarConnectorId connectorId,
+            PulsarConnectorConfig pulsarConnectorConfig,
+            PulsarDispatchingRowDecoderFactory decoderFactory,
+            PulsarAdmin pulsarAdmin)
+    {
+        this.decoderFactory = decoderFactory;
+        this.connectorId = requireNonNull(connectorId, "connectorId is null").toString();
+        this.pulsarConnectorConfig = pulsarConnectorConfig;
+        this.hideInternalColumns = pulsarConnectorConfig.isHideInternalColumns();
+        this.pulsarAdmin = pulsarAdmin;
+    }
+
+    @Override
+    public List<String> listSchemaNames(ConnectorSession session)
+    {
+        List<String> trinoSchemas = new LinkedList<>();
+        try {
+            List<String> tenants = pulsarAdmin.tenants().getTenants();
+            for (String tenant : tenants) {
+                trinoSchemas.addAll(pulsarAdmin.namespaces().getNamespaces(tenant).stream().map(namespace ->
+                        PulsarConnectorUtils.rewriteNamespaceDelimiterIfNeeded(namespace, pulsarConnectorConfig)).collect(Collectors.toList()));
+            }
+        }
+        catch (PulsarAdminException e) {
+            if (e.getStatusCode() == 401) {
+                throw new TrinoException(QUERY_REJECTED, "Failed to get schemas from pulsar: Unauthorized");
+            }
+            throw new RuntimeException("Failed to get schemas from pulsar: " + Throwables.getRootCause(e).getLocalizedMessage(), e);
+        }
+        return trinoSchemas;
+    }
+
+    @Override
+    public ConnectorTableHandle getTableHandle(ConnectorSession session,
+                                               SchemaTableName tableName)
+    {
+        TopicName topicName = getMatchedTopicName(tableName);
+        return topicName == null ? null : new PulsarTableHandle(
+                connectorId,
+                tableName.getSchemaName(),
+                tableName.getTableName(),
+                topicName.getLocalName());
+    }
+
+    @Override
+    public List<ConnectorTableLayoutResult> getTableLayouts(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            Constraint constraint,
+            Optional<Set<ColumnHandle>> desiredColumns)
+    {
+        PulsarTableHandle handle = PulsarHandleResolver.convertTableHandle(table);
+        ConnectorTableLayout layout = new ConnectorTableLayout(
+                new PulsarTableLayoutHandle(handle, constraint.getSummary()));
+        return ImmutableList.of(new ConnectorTableLayoutResult(layout, constraint.getSummary()));
+    }
+
+    @Override
+    public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTableLayoutHandle handle)
```

Comment:
Table layouts are going to be removed.  Can you please return `true` for `io.trino.spi.connector.ConnectorMetadata#usesLegacyTableLayouts`

Reply:
updated.